src/main/kotlin/org/jetbrains/kotlin/jupyter/config.kt

```diff
@@ -27,16 +53,99 @@ class LibraryDefinition(val dependencies: List<String>,
                         val renderers: List<TypeRenderer>,
                         val link: String?)
 
-data class ResolverConfig(val repositories: List<RepositoryCoordinates>, val libraries: Map<String, LibraryDefinition>)
+data class ResolverConfig(val repositories: List<RepositoryCoordinates>,
+                          val libraries: Map<String, LibraryDefinition>)
 
-data class KernelConfig(
-        val ports: Array<Int>,
-        val transport: String,
-        val signatureScheme: String,
-        val signatureKey: String,
-        val pollingIntervalMillis: Long = 100,
-        val scriptClasspath: List<File> = emptyList(),
-        val resolverConfig: ResolverConfig?
+fun readJson(path: String) =
+        Parser.default().parse(path) as JsonObject
+
+fun JSONObject.toJsonObject() = Parser.default().parse(StringReader(toString())) as JsonObject
+
+fun <T> catchAll(body: () -> T): T? = try {
+    body()
+} catch (e: Exception) {
+    null
+}
+
+fun parseLibraryArgument(str: String): Variable {
+    val eq = str.indexOf('=')
+    return if (eq == -1) Variable(str.trim(), null)
+    else Variable(str.substring(0, eq).trim(), str.substring(eq + 1).trim())
+}
+
+fun parseLibraryName(str: String): Pair<String, List<Variable>> {
+    val brackets = str.indexOf('(')
+    if (brackets == -1) return str.trim() to emptyList()
+    val name = str.substring(0, brackets).trim()
+    val args = str.substring(brackets + 1, str.indexOf(')', brackets))
+            .split(',')
+            .map(::parseLibraryArgument)
+    return name to args
+}
+
+fun readLibraries(basePath: String? = null, filter: (File) -> Boolean = { true }): List<Pair<String, JsonObject>> {
+    val parser = Parser.default()
+    return File(basePath, LibrariesDir)
+            .listFiles()?.filter { it.extension == "json" && filter(it) }
+            ?.map {
+                log.info("Loading '${it.nameWithoutExtension}' descriptor from '${it.canonicalPath}'")
+                it.nameWithoutExtension to parser.parse(it.canonicalPath) as JsonObject
+            }
+            .orEmpty()
+}
+
+fun getLibrariesJsons(homeDir: String): Map<String, JsonObject> {
+
+    val librariesMap = readLibraries(LocalSettingsPath).toMap().orEmpty().toMutableMap()
+
+    val address = GitHubApiPrefix + LibrariesDir
+    val response = catchAll { khttp.get(address) }
+    if (response != null && response.statusCode == 200) {
+        response.jsonArray.forEach {
+            val o = it as JSONObject
+            val filename = o["name"] as String
+            if (filename.endsWith(".json")) {
+                val libName = filename.substring(0, filename.length - 5)
+                if (!librariesMap.containsKey(libName)) {
```

Comment:
Want to encapsulate this check somehow, because the same check is done below... But it's optional.

Reply:
done